### ISSUE TITLE
[java] PreserveStackTraceRule: false positive when using builder pattern

### ISSUE THREAD

**Issue by dreniers** [pmd/pmd#422]
**Rule Set:**

rulesets/java/design.xml/PreserveStackTrace

**Description:**

The PreserveStackTrace tries to detect when stack trace is preserved when the exception is wrapped as an inner exception, but fails when using a builder-like pattern, see the third throw in the code snippet below.

**Code Sample demonstrating the issue:**

```
public final class Main {

    public void execute() {
        final String dirname = "c:/temp";

        try {
            FileUtils.deleteDirectoryContent(Paths.get(dirname));
        } catch (final IOException e) {
            throw new ApplicationException("Error"); // throw 1: true positive
        }

        try {
            FileUtils.deleteDirectoryContent(Paths.get(dirname));
        } catch (final IOException e) {
            throw new ApplicationException("Error", e); // throw 2: true negative
        }

        try {
            FileUtils.deleteDirectoryContent(Paths.get(dirname));
        } catch (final IOException e) {
            throw uncheckedException(ErrorCodeCommon.DIRECTORY_NOT_FOUND) // throw 3: false positive
                    .withField("dirname", dirname)
                    .causedBy(e) 
                    .build(); 
        }
    }
}
```

**Running PMD through:** *CLI*

```
pmd.bat -d ClearAdditionalDebugData.java -f xml -R rulesets/java/design.xml/PreserveStackTrace
```
Outputs:
```
<?xml version="1.0" encoding="UTF-8"?>
<pmd version="5.6.1" timestamp="2017-06-02T16:12:47.550">
<file name="E:\tickets\18522\bin\ClearAdditionalDebugData.java">
<violation beginline="9" endline="9" begincolumn="13" endcolumn="52" rule="PreserveStackTrace" ruleset="Design" class="Main" method="execute" externalInfoUrl="h
ttps://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/design.html#PreserveStackTrace" priority="3">
New exception is thrown in catch block, original stack trace may be lost
</violation>
<violation beginline="21" endline="22" begincolumn="13" endcolumn="29" rule="PreserveStackTrace" ruleset="Design" class="Main" method="execute" externalInfoUrl=
"https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/design.html#PreserveStackTrace" priority="3">
New exception is thrown in catch block, original stack trace may be lost
</violation>
</file>
</pmd>
```

**Comment by jsotuyod** [comment]
This is a hard one, since it requires multi-class data flow analysis to be able to get it right.

Definitely an enhancement, but not likely we will be able to tackle this anytime soon.

**Comment by dreniers** [comment]
The rule currently does not do an extensive data flow analysis either. It simply checks whether the first ASTArgumentList in the ASTThrowStatement contains an ASTName that refers to the caught exception. This means that `throw uncheckedException(e)` does not produce the violation whereas `uncheckedException(ErrorCodeCommon.DIRECTORY_NOT_FOUND).causedBy(e).build()` does, which does not feel as a meaningful distinction to me.

I propose that instead of checking the ASTName of the first ASTArgumentList, we check all ASTNames that occur in the ASTThrowStatement. I am going to submit a pull request for this.

**Comment by jsotuyod** [comment]
@dreniers interesting alternative, I'll check it.
